### ISSUE TITLE
Machines: Improve virtual network layout

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#11253]
The layout in https://github.com/cockpit-project/cockpit/pull/11204 can be improved.

Currently, it looks like:

![example VM network layout](https://user-images.githubusercontent.com/42733240/53335873-ef58d080-38fc-11e9-8318-6ea9f7eb4755.png)

There's also a QoS section, like in this old (and problematic, due to vertical stretching) example:

![](https://user-images.githubusercontent.com/42733240/53095656-b0ea9c80-351d-11e9-88f4-22f0a0aa2db2.png)

**Comment by garrett** [comment]
Quoting myself @ https://github.com/cockpit-project/cockpit/pull/11204#issuecomment-467378427:

> It looks quite odd with the IPv4 & IPv6 blocks right-aligned with a bunch of white-space on the left and between the blocks.
>
> We could have the blocks wrap around, depending on space... or it could always be like this:
>
> General | QoS _(optional)_
> -- | --
> **IPv4** | **IPv6**
>
> ...as both General and QoS are pretty small, whereas IPv4 & IPv6 are large. Downside of this approach is if there's only General and IPv4. In that case, we might be able to somehow work it out so that General and IPv4 are side by side.

**Comment by garrett** [comment]
I suppose one method to make sure everything's in order and space is used properly could be to group and subgroup elements, so that they do a bit of sliding / wrapping on their own.

It would be like these text-based visualizations:

All items, decently wide viewport:

```
[
  [ General | QoS ]
  [ IPv4 | IPv6 ]
]
```

With a wider screen and/or QoS missing, it'd slide like this:
```
[
  [ General | ] [ IPv4 | IPv6 ]
]
```

Same exact layout on a narrower screen would just drop QoS and keep IPv4 & IPv6 together:
```
[
  [ General | ]
  [ IPv4 | IPv6 ]
]
```

But dropping another item in the same screen width would let it slide like so: 
```
[
  [ General | ] [ IPv4 |]
]
```

If the viewport is quite narrow, it might end up like:
```
[
  [ General
  | QoS ]
  [ IPv4
  | IPv6 ]
]
```

There might be cases were the viewport is pretty narrow, but General and QoS are so small it may look like this too:
```
[
  [ General | QoS ]
  [ IPv4
  | IPv6 ]
]
```

However, all of the permutations above have the following in common:
1. It's in the same source order and visual order.
2. The layout slides around for a best fit based on content size within viewport size.
3. It attempts to keep associated items together. (IPv4 & IPv6)

It should be able to be used without breakpoints, with everything just adapting. We'd probably want to use nested grids for it, so we could use `fr` units.

**Comment by garrett** [comment]
It's probably possible (and may make even more sense) to implement this like so:

```
[
  [ General ] [ QoS ]
  [ IPv4 | IPv6 ]
]
```

That is, have 3 top-level items, with only IPv4 & IPv6 being a inside a group, ensuring they're shown together within permitting space.

**Comment by garrett** [comment]
So, of course, the "big" problem with this layout idea is that the labels inside the components won't line up with labels from other components. :cry: 

The solution is to use [CSS subgrid](https://www.w3.org/TR/css-grid-2/#subgrids). ([Explanation with examples](https://dev.to/kenbellows/why-we-need-css-subgrid-53mh).)

However, browser support isn't available... yet.

A fallback approach is to use some local CSS to set min-width on the labels to the longest label. Of course, this won't work reliably across languages. However, if the label exceeds the min-width, then it would push all the labels in the ct-form-layout wider... but not in the other layout blocks. It would be a display thing, but nothing would really break.

Whenever subgrid finally lands across browsers, it would be possible to move the min-width fallback hack into a `@supports (not (grid-template-columns: subgrid)) {}` block.

However, this is still better than anything else we can do at the moment.

---

_**TL;DR**: Visual optimization would be to set a min-width for when the boxes stack, as a best-effort attempt to align labels._

**Comment by garrett** [comment]
Stepping back, the goal is to have layouts like these:


General | QoS | IPv4 | IPv6
-- | -- | -- | --


General | IPv4 | IPv6
-- | -- | --



General |  
-- | --
**IPv4** | **IPv6**



General | IPv4
-- | --

General |
-- |
**QoS** |
**IPv4** |
**IPv6** |





But _**not**_ this:

**General** | **IPv4**
-- | --
**IPv6** |
